File: 20-valid-parentheses/20-valid-parentheses.py
class Solution:
    def isValid(self, st: str) -> bool:
        res = []
        for s in st:
            if s in ['(','{','[']:
                res.append(s)
            elif len(res)==0:
                return False
            elif s==')' and res[-1]=='(':
                res.pop()
            elif s=='}' and res[-1]=='{':
                res.pop()
            elif s==']' and res[-1]=='[':
                res.pop()
            else:
                return False
        if len(res)==0:
            return True
        return False
        

File: 20-valid-parentheses/20-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        
        stack<char> st;
        
        unordered_map<char,int> mp;
        mp['('] = 1;    mp[')'] = 10;
        mp['{'] = 2;    mp['}'] = 20;
        mp['['] = 3;    mp[']'] = 30;
        
        for(auto c: s){
            if(mp[c] <= 3)
                st.push(c);
            else{
                if(st.empty())
                    return false;
                if(mp[st.top()]*10 == mp[c])
                    st.pop();
                else
                    return false;
            }
        }
        if(st.empty())
            return true;
        else
            return false;
    }
};